hooks/useTokenPayment.js

```diff
@@ -9,104 +9,85 @@ import useSWR from 'swr';
+  const totalNightsContractForm = Array.from(
+    { length: totalNights },
+    (_, i) => [bookingYear, dayjs(startDate).add(i, 'day').dayOfYear()],
+  );
+  const { data: bookedNights, mutate: updateBookedNights } = useSWR(
+    [
+      BLOCKCHAIN_DAO_DIAMOND_ADDRESS,
+      'getAccommodationBookings',
+      account,
+      bookingYear,
+    ],
     {
       fetcher: fetcher(library, BLOCKCHAIN_DIAMOND_ABI),
     },
   );
 
-  const tokensToStake =
-    balanceLocked &&
-    BigNumber.from(value)
-      .mul(BigNumber.from(10).pow(BLOCKCHAIN_DAO_TOKEN.decimals))
-      .sub(balanceLocked);
+  const isStakingConfirmedOnBlockchain = useMemo(() => {
+    if (!bookedNights) return false;
+    const areAllNightsBooked = totalNightsContractForm.every((night) => {
+      const [year, dayOfYear] = night;
+      return bookedNights.some(
+        (bookedNight) =>
+          bookedNight.year === year && bookedNight.dayOfYear === dayOfYear,
+      );
+    });
+
+    return areAllNightsBooked;
+  }, [bookedNights, totalNightsContractForm]);
 
   const stakeTokens = async () => {
-    if (!library || !account || !tokensToStake) {
+    if (!library || !account) {
       return;
     }
 
-    const DAOTokenContract = new Contract(
-      BLOCKCHAIN_DAO_TOKEN.address,
-      BLOCKCHAIN_DAO_TOKEN_ABI,
-      library.getUncheckedSigner(),
-    );
-
     const Diamond = new Contract(
       BLOCKCHAIN_DAO_DIAMOND_ADDRESS,
       BLOCKCHAIN_DIAMOND_ABI,
       library.getUncheckedSigner(),
     );
 
-    setPending(true);
-    //Approve the contract to spend tokens
-    if (tokensToStake > 0) {
-      try {
-        const tx1 = await DAOTokenContract.approve(
-          BLOCKCHAIN_DAO_DIAMOND_ADDRESS,
-          tokensToStake,
-        );
-        setPendingTransactions([...pendingTransactions, tx1.hash]);
-        await tx1.wait();
-        console.log(`${tx1.hash} mined`);
-        setPendingTransactions((pendingTransactions) =>
-          pendingTransactions.filter((h) => h !== tx1.hash),
-        );
-      } catch (error) {
-        console.error('useTokenPayment step 1', error);
-        setPending(false);
-        return {
-          error: { ...error, message: 'Error on step 1 ' + error.message },
-          success: null,
-        };
-      }
-    }
-
-    //Now we can book the nights
     try {
       setPending(true);
-      const tx3 = await Diamond.bookAccommodation(nights);
+      const pricePerNightBigNum = BigNumber.from(dailyValue).mul(
+        BigNumber.from(10).pow(BLOCKCHAIN_DAO_TOKEN.decimals),
+      );
+      const tx3 = await Diamond.bookAccommodation(
+        totalNightsContractForm,
+        pricePerNightBigNum,
+      );
+
       setPendingTransactions([...pendingTransactions, tx3.hash]);
       await tx3.wait();
-      console.log(`${tx3.hash} mined`);
       setPendingTransactions((pendingTransactions) =>
         pendingTransactions.filter((h) => h !== tx3.hash),
       );
-      setPending(false);
+      updateBookedNights();
```

Comment:
Also update tokenBalance for the user might want to look at its token balance here